cas-server-core/src/main/java/org/jasig/cas/services/DefaultRegisteredServiceAuthorizationStrategy.java

```diff
@@ -0,0 +1,258 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Verify presence of service required attributes.
+     * <ul>
+     *     <li>If no required attributes are specified, authz is granted.</li>
+     *     <li>If ALL required attributes must be present, and the principal contains all and there is
+     *     at least one attribute value that matches the required, authz is granted.</li>
+     *     <li>If ALL required attributes don't have to be present, and there is at least
+     *     one principal attribute present whose value matches the required, authz is granted.</li>
+     *     <li>Otherwise, access is denied</li>
+     * </ul>
+     * Note that comparison of principal/required attributes is case-sensitive. Exact matches are required
+     * for any individual attribute value.
+     */
+    @Override
+    public boolean isServiceAccessAuthorizedForPrincipal(final Map<String, Object> principalAttributes, final Service service) {
+        if (getRequiredAttributes().isEmpty()) {
+            logger.debug("No required attributes are specified");
+            return true;
+        }
+        if (principalAttributes.isEmpty()) {
+            logger.warn("No principal attributes are found to satisfy attribute requirements for [{}]", service.getId());
+            return false;
+        }
+
+        if (principalAttributes.size() < getRequiredAttributes().size()) {
+            logger.warn("The size of the principal attributes that are [{}] does not match requirements, "
+                    + "which means the principal is not carrying enough data to grant authorization",
+                    principalAttributes);
+            return false;
+        }
+
+        logger.debug("These required attributes [{}] are examined against [{}] before service [{}] can proceed.",
+                getRequiredAttributes(), principalAttributes, service.getId());
+
+        final Sets.SetView<String> difference = Sets.intersection(getRequiredAttributes().keySet(), principalAttributes.keySet());
+        final Set<String> copy = difference.immutableCopy();
+
+        if (isRequireAllAttributes() && copy.size() < getRequiredAttributes().size()) {
+            logger.warn("Not all required attributes are available to the principal");
+            return false;
+        }
+
+        for (final String key : copy) {
+            final Set<?> requiredValues = getRequiredAttributes().get(key);
+            Set<?> availableValues = null;
+
+            final Object objVal = principalAttributes.get(key);
+            if (objVal instanceof Collection) {
+                final Collection valCol = (Collection) objVal;
+                availableValues = Sets.newHashSet(valCol.toArray());
+            } else {
+                availableValues = Collections.singleton(objVal);
+            }
+
+            final Sets.SetView<?> differenceInValues = Sets.intersection(availableValues, requiredValues);
+            if (differenceInValues.size() > 0) {
```

Comment:
So even if I must have all attributes, I only need one value in one attribute to be authorized. I would have expected to be authorized if I have one value for each requried attribute.